prov/opx/src/fi_opx_init.c

```
@@ -0,0 +1,481 @@
+	 * has an opened instance of the named domain, this (fi_domain_attr::domain)
+	 * will reference the first opened instance. If no instance has been
+	 * opened, this field will be NULL.
+	 */
+
+	ret = fi_opx_choose_domain(caps, fi->domain_attr,
+		(hints)?(hints->domain_attr):NULL);
+	if (ret) {
+		FI_LOG(fi_opx_global.prov, FI_LOG_DEBUG, FI_LOG_FABRIC,
+				"cannot find appropriate domain\n");
+		goto err;
+	}
+
+	memcpy(fi->ep_attr, fi_opx_global.default_ep_attr, sizeof(*fi->ep_attr));
+	if (hints && hints->ep_attr) {
+		/* adjust parameters down from what requested if required */
+		fi->ep_attr->type	= hints->ep_attr->type;
+		if (hints->ep_attr->max_msg_size > 0 &&
+			hints->ep_attr->max_msg_size <= fi_opx_global.default_ep_attr->max_msg_size)
+				fi->ep_attr->max_msg_size = hints->ep_attr->max_msg_size;
+
+		if (0 != hints->ep_attr->tx_ctx_cnt && hints->ep_attr->tx_ctx_cnt <= fi->ep_attr->tx_ctx_cnt)
+			fi->ep_attr->tx_ctx_cnt = hints->ep_attr->tx_ctx_cnt;	/* TODO - check */
+
+		if (0 != hints->ep_attr->rx_ctx_cnt && hints->ep_attr->rx_ctx_cnt <= fi->ep_attr->rx_ctx_cnt)
+			fi->ep_attr->rx_ctx_cnt = hints->ep_attr->rx_ctx_cnt;	/* TODO - check */
+	}
+
+	return 0;
+err:
+	if (fi->domain_attr->name) free(fi->domain_attr->name);
+	if (fi->fabric_attr->name) free(fi->fabric_attr->name);
+	if (fi->fabric_attr->prov_name) free(fi->fabric_attr->prov_name);
+	errno = FI_ENODATA;
+	return -errno;
+}
+
+struct fi_opx_global_data fi_opx_global;
+
+static int fi_opx_getinfo(uint32_t version, const char *node,
+		const char *service, uint64_t flags,
+		const struct fi_info *hints, struct fi_info **info)
+{
+	int ret;
+	struct fi_info *fi;
+
+	if (!fi_opx_count) {
+		errno = FI_ENODATA;
+		return -errno;
+	}
+
+	if (hints) {
+		ret = fi_opx_check_info(hints);
+		if (ret) {
+			return ret;
+		}
+		if (!(fi = fi_allocinfo())) {
+			return -FI_ENOMEM;
+		}
+		if (fi_opx_fillinfo(fi, node, service,
+					hints, flags)) {
+			return -errno;
```

It would be better to return an error value from fi_opx_fillinfo rather than relying on errno.

---

Noted, but it doesn't seem to matter. I'm seeing this effect even when fi_opx_getinfo returns success.

---

I will address this in the next push.